cmd/pod-scaler/consumer.go

```diff
@@ -0,0 +1,130 @@
+package main
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/interrupts"
+
+	pod_scaler "github.com/openshift/ci-tools/pkg/pod-scaler"
+)
+
+func newReloader(name string, cache cache) *cacheReloader {
+	reloader := &cacheReloader{
+		name:  name,
+		cache: cache,
+		logger: logrus.WithFields(logrus.Fields{
+			"component": "reloader",
+			"metric":    name,
+		}),
+		lock: &sync.RWMutex{},
+	}
+	interrupts.TickLiteral(reloader.reload, 10*time.Minute)
+	return reloader
+}
+
+type cacheReloader struct {
+	name   string
+	cache  cache
+	logger *logrus.Entry
+
+	lock        *sync.RWMutex
+	lastUpdated time.Time
+	lastLoaded  *pod_scaler.CachedQuery
+	subscribers []chan<- interface{}
+}
+
+func (c *cacheReloader) subscribe(out chan<- interface{}) {
+	c.lock.Lock()
+	c.subscribers = append(c.subscribers, out)
+	c.lock.Unlock()
+}
+
+func (c *cacheReloader) reload() {
+	// technically this can race as we read the attribute and data from the handle at
+	// different times, but there doesn't seem to be an atomic call to GCS for that anyway
+	lastUpdated, err := lastUpdated(c.cache, c.name)
+	if err != nil {
+		c.logger.WithError(err).Warn("Failed to query for last cache update time, won't reload this tick.")
+		return
+	}
+	c.lock.RLock()
+	lastSeen := c.lastUpdated
+	c.lock.RUnlock()
+	logger := c.logger.WithFields(logrus.Fields{
+		"last_update_seen": lastSeen.Format(time.RFC3339),
+		"last_update":      lastUpdated.Format(time.RFC3339),
+	})
+
+	if lastUpdated == lastSeen {
+		logger.Debug("Last updated time on cloud artifacts matches our last load, won't reload this tick.")
+		return
+	}
+	logger.Debug("Newer update available in cloud storage, reloading data.")
+
+	data, err := loadCache(c.cache, c.name, c.logger)
+	if err != nil {
+		logger.WithError(err).Warn("Failed to read cached data, won't reload this tick.")
+		return
+	}
+	c.lock.Lock()
+	c.lastUpdated = lastUpdated
+	c.lastLoaded = data
+	for _, subscriber := range c.subscribers {
+		subscriber <- struct{}{}
+	}
+	c.lock.Unlock()
+	logger.Debug("Newer update loaded.")
+}
+
+func (c *cacheReloader) data() *pod_scaler.CachedQuery {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	return c.lastLoaded
```

Comment:
nit: I'm slightly triggered that you sometimes defer the unlock and sometimes you call it directly (`subscribe`), makes skimming harder when there's no pattern